Truncate tags command clipboard text within 32k string limit

sciwhiz12 opened the pull request:
This PR fixes #1543 by truncating the number of lines in the copy-to-clipboard text of the `/neoforge tags` command to as much entries as can fit within the (artificial) limit of 32.6k characters for elements (plus some allowance for the explanatory text for the truncation).

Tested by modifying the `list` subcommand temporarily to have 30,000+ dummy entries (adding to the count and using `Stream.concat`), and it works.

A previous version of this PR artifically constrained the number of entries to 1,000 entries (approximately the number of 32-character-long entries that fits into 32,767--the limit of strings in `FriendlyByteBuf`--and rounded down the nearest hundreds). At @Technici4n's prompting (in Discord), I've changed the strategy to instead fit as much of the entries as possible by counting their lengths as they're added to the text, and cutting off when the text is about to reach a limit, which is current 32,600 characters. That leaves about ~167 characters room for the explanatory note for the truncation, which should be plenty.

---

We could perhaps consider adding a new subcommand to `/neoforge dump` which exports the list of tags for registries (or elements in a tag, or tags of an element). Should that be added, we may also remove the functionality of the copy-on-clipboard for `/neoforge tags`, and perhaps even have it instead give the appropriate `/neoforge dump` command to the user with arguments filled in.

sciwhiz12 committed:
Truncate tags command clipboard text to 1000 entries

Fixes #1543

HenryLoenwind commented:
One thought: Does this really need to run as a server command? Those registries are synced to the client, aren't they?

lukebemish commented:
Not _necessarily_ -- afaik that command dumps tags for any registry, including server only ones? Registries can have tags and not be synced.

sciwhiz12 committed:
Merge latest 1.21.x

sciwhiz12 committed:
Merge latest 1.21.x

sciwhiz12 committed:
Change truncation strategy to fit as much text

This may be less elegant to look at and read, but it fits more into the
32.6k limit we (artificially) set for the clipboard text by actually
measuring the text as we go add more elements, until no more can fit
within the limit.

sciwhiz12 commented:
Updated to use a new truncation strategy that fits as much text as possible into the clipboard text, rather than constraining the amount of elements based on a precalculated heuristic on element lengths. See updated PR description for details.

Technici4n commented on src/main/java/net/neoforged/neoforge/server/command/TagsCommand.java:
```diff
@@ -171,16 +175,45 @@ private static MutableComponent createMessage(final MutableComponent header,
+
+            if (reachedLimit) {
+                // Almost went over the limit; add additional info to clipboard text
+                clipboardText = "(Too many entries to fit in clipboard, showing only first " + countedLines + " entries...)" + '\n'
```
Shouldn't there be a `\n` at the beginning since you remove it just above?

sciwhiz12 replied:
What do you mean? The clipboard text builder will contain a string that doesn't end in a newline, and in the case of the limit being reached, the starting truncation notice and the ending truncation notice are both separated by a newline from the main (elements) text.

Technici4n replied:
Ah right this looks correct.
